File: NX10.0.0.24/UGOPEN/SampleNXOpenApplications/.NET/DrawingCycle/drawing_cycle.vb
' NX 3.0.0.10
' Journal created by geolekas on Fri Mar 05 15:18:51 2004 Eastern Standard Time
Option Strict Off
Imports System
Imports NXOpen
Imports NXOpen.Drawings
Imports NXOpen.Annotations
Imports System.Text

Module NXJournal
Sub Main

' ----------------------------------------------
' This example opens up a part, cycles through
' all the sheets in the part. For each sheet, it updates
' all the views on it.  It then places a note in the upper left corner
' that has the current date on it.   And then it popups a message box.   
' the sheet.  (A programmer could add instructions to plot the sheet"
' And then it deletes the note just added to the sheet.  
'
' NOTE: This example is setup to place a note correctly for 
' an E size drawing.  
'
' If you do not want to delete the note that is created,  
' change the constant value of DO_NOT_DELETE_NOTE to be 1.  
' ----------------------------------------------
#Const DO_NOT_DELETE_NOTE = 0


Dim theSession As Session = Session.GetSession()


' ----------------------------------------------
'   Menu: File->Open...
' Fill in part location on the OpenDisplay line.
' ----------------------------------------------
Dim part1 As Part

part1 = theSession.Parts.Work



Dim drawingSheets As DrawingSheet()
Dim mySheet As DrawingSheet
drawingSheets = part1.DrawingSheets.ToArray

Dim dfviews() As DraftingView
Dim dfview As DraftingView

' ----------------------------------------------
' Here we loop through all the sheets in the part.
' For each sheet, we open the sheet, update all views in it,
' And then pop-up a message box listing the 
' views on the sheet.  This could be modified to plot out
' the sheets instead.
' ----------------------------------------------

For Each mySheet In drawingSheets ' Iterate through elements.

    Dim session_UndoMarkId1 As Session.UndoMarkId
    session_UndoMarkId1 = theSession.SetUndoMark(Session.MarkVisibility.Visible, "Open Drawing")
    mySheet.Open()


    ' ----------------------------------------------
    '   Menu: Edit->View->Update Views...
    ' ----------------------------------------------
    Dim session_UndoMarkId2 As Session.UndoMarkId
    session_UndoMarkId2 = theSession.SetUndoMark(Session.MarkVisibility.Visible, "Drawing View Update")
    'There are a couple of  ways to update all the views in the part
    ' we will use the update all views in sheet method, vs. building up
    ' an array of views to be used for update.
    theSession.Parts.Work.DraftingViews.UpdateViews(DraftingViewCollection.ViewUpdateOption.All,mySheet)


    ' ----------------------------------------------
    '   Menu: Insert->Annotation...
    ' ----------------------------------------------
    Dim session_UndoMarkId3 As Session.UndoMarkId
    session_UndoMarkId3 = theSession.SetUndoMark(Session.MarkVisibility.Invisible, "Create Annotation")

    Dim letteringPreferences1 As LetteringPreferences
    letteringPreferences1 = theSession.Parts.Work.Annotations.Preferences.GetLetteringPreferences()

    Dim userSymbolPreferences1 As UserSymbolPreferences
    userSymbolPreferences1 = theSession.Parts.Work.Annotations.NewUserSymbolPreferences(UserSymbolPreferences.SizeType.ScaleAspectRatio, 1, 1)

    Dim TimeNote As DateTime
    TimeNote = System.DateTime.Now()
    Dim TimeNoteString as String() = TimeNote.GetDateTimeFormats()

    Dim stringArray2(1) As String
    stringArray2(0) = "<c3.3.25> Printed on:"
    stringArray2(1) = TimeNoteString(0) + "<c>"
    Dim point3d1 As Point3d = New Point3d(40.8263111888112, 33.145541958042, 0)

    Dim note1 As Note
    note1 = theSession.Parts.Work.Annotations.CreateNote(stringArray2, point3d1, AxisOrientation.Horizontal, letteringPreferences1, userSymbolPreferences1)

    theSession.SetUndoMarkVisibility(session_UndoMarkId3, "Create Annotation", Session.MarkVisibility.Visible)


    letteringPreferences1.Dispose()

    userSymbolPreferences1.Dispose()

    ' ----------------------------------------------
    '   Print will go here
    ' ----------------------------------------------
    dfviews = mySheet.GetDraftingViews()
    Dim sheet_name As String
    Dim view_names As StringBuilder = new StringBuilder()

    sheet_name = "Printing Sheet: " + mySheet.Name 
    
    view_names.append("Sheet contains the following views: ")

    For Each dfview in dfviews
       view_names.append(dfview.Name +" ")
    Next


    Windows.Forms.MessageBox.Show(view_names.ToString(), sheet_name)
#If DO_NOT_DELETE_NOTE = 0 Then

    ' ----------------------------------------------
    '   Menu: Edit->Delete...
    ' ----------------------------------------------
    theSession.UpdateManager.ClearErrorList()

    Dim session_UndoMarkId5 As Session.UndoMarkId
    session_UndoMarkId5 = theSession.SetUndoMark(Session.MarkVisibility.Visible, "Delete")

    Dim integer4 As Integer
    integer4 = theSession.UpdateManager.AddToDeleteList(note1)

    Dim integer5 As Integer
    integer5 = theSession.UpdateManager.DoUpdate(session_UndoMarkId5)
    
#End If
Next


End Sub
End Module

File: NX1953_NX1957_NX1961_NX1965_NX1969_NX1973/UGOPEN/SampleNXOpenApplications/C++/DrawingCycle/DrawingCycle.cpp
//===============================================================================
//
//                   Copyright (c) 2010 Siemens PLM Solutions
//                     Unpublished - All rights reserved
//========================================================================================
//File description: NX/Open example to Cycle the Drawing sheets and Drawing Note Creation
//========================================================================================
//

/* Include files */
#   include <sstream>
#   include <iostream>
using std::ostringstream;
using std::endl;    
using std::ends;
using std::cerr;


#include <uf.h>
#include <uf_ui.h>
#include <uf_defs.h>
#include <string.h>
#include <sstream>
#include <iostream>
#include <NXOpen/NXException.hxx>
#include <NXOpen/Drawings_DraftingViewCollection.hxx>
#include <NXOpen/Drawings_DrawingSheetCollection.hxx>
#include <NXOpen/Drawings_SectionView.hxx>
#include <NXOpen/Drawings_UpdateViewsBuilder.hxx>
#include <NXOpen/Annotations_AnnotationManager.hxx>
#include <NXOpen/Annotations_Note.hxx>
#include <NXOpen/NXObject.hxx>
#include <NXOpen/PartCollection.hxx>
#include <NXOpen/UI.hxx>
#include <NXOpen/NXMessageBox.hxx>
#include <NXOpen/Preferences_AnnotationPreferences.hxx>

using namespace std;
using namespace NXOpen;
using namespace NXOpen::Annotations;


/*****************************************************************************
 **  Activation Methods
 *****************************************************************************/
/*  Explicit Activation
 **      This entry point is used to activate the application explicitly, as in
 **      "File->Execute UG/Open->User Function..." */
extern DllExport void ufusr( char *parm, int *returnCode, int rlen )
{
    /* Initialize the API environment */
    int errorcode = UF_initialize();
    
    /* Add your application code here */
    Session *theSession = Session::GetSession();
    Part *workPart(theSession->Parts()->Work());
    Part *displayPart(theSession->Parts()->Display());
    
    
    //Getting Drawing sheets into collection of the workpart
    Drawings::DrawingSheetCollection *drawingsheets = workPart->DrawingSheets();
    Drawings::DrawingSheetCollection::iterator it;
    Drawings::DrawingSheet *mysheet;
    
    /*----------------------------------------------
     Here we loop through all the sheets in the part.
     For each sheet, we open the sheet, update all views in it,
     And then pop-up a message box listing the 
     views on the sheet.  This could be modified to plot out
     the sheets instead.
     ----------------------------------------------*/
    for(it=drawingsheets->begin();it!=drawingsheets->end();it++)
    {
        
        mysheet = *it;
        mysheet->Open();
        // ----------------------------------------------
        //   Menu: Edit->View->Update Views...
        // ----------------------------------------------
        theSession->Parts()->Work()->DraftingViews()->UpdateViews(Drawings::DraftingViewCollection::ViewUpdateOptionAll,mysheet);
        
        Annotations::LetteringPreferences *letteringPreferences1;
        letteringPreferences1 = workPart->Annotations()->Preferences()->GetLetteringPreferences();
        
        
        //Setting Lettering Preferences
        Annotations::Lettering Annotations_Lettering1;
        Annotations_Lettering1.Size = 0.35;
        Annotations_Lettering1.CharacterSpaceFactor = 1;
        Annotations_Lettering1.AspectRatio = 1.0;
        Annotations_Lettering1.LineSpaceFactor = 1.0;
        Annotations_Lettering1.Cfw.Color = 2;
        Annotations_Lettering1.Cfw.Font = 1;
        Annotations_Lettering1.Cfw.Width = Annotations::LineWidthThin;
        letteringPreferences1->SetGeneralText(Annotations_Lettering1);
        
        workPart->Annotations()->Preferences()->SetLetteringPreferences(letteringPreferences1);
        // ----------------------------------------------
        //   Menu: Insert->Annotation...
        // ----------------------------------------------
        Annotations::UserSymbolPreferences *userSymbolPreferences1 = theSession->Parts()->Work()->Annotations()->NewUserSymbolPreferences(UserSymbolPreferences::SizeTypeScaleAspectRatio,1,1);
        UF_system_info_t info;
        UF_ask_system_info(&info);
        std::ostringstream buffer; 
        buffer << info.date_buf; 
        NXOpen::NXString currDate = buffer.str(); 
        string date = currDate.GetText();
        size_t pos1 = date.find_first_of(" ");
        date = date.substr(0,pos1);
        currDate = date;
        
        
        std::vector<NXString> stringArray1(1);
        stringArray1[0]=currDate.GetText();
        std::vector<NXString> stringArray2(2);        
        stringArray2[0]="Printed On";
        stringArray2[1]=currDate.GetText();
        
        
        NXOpen::Point3d point3d1(30.643873924,1.999761255 ,0);
        
        //NULL INPUT TO BE CHANGED
        //The below method will create a note.
        
        NXOpen::Annotations::Note *note1 = theSession->Parts()->Work()->Annotations()->CreateNote(stringArray2,point3d1,AxisOrientationHorizontal,letteringPreferences1,userSymbolPreferences1);
        delete userSymbolPreferences1;
        delete letteringPreferences1;
        
        // ----------------------------------------------
        //   Printing Drafting views
        // ----------------------------------------------
        std::vector<NXOpen::Drawings::DraftingView *> dfviews = mysheet->GetDraftingViews();
        NXOpen::NXString sheet_name = "Printing Sheet: " + mysheet->Name();
        std::vector<NXOpen::NXString >sheet_names(1);
        UI *theUI = UI::GetUI();         
        
        //Collecting View names and copying to String
        for(int i=0;i!=dfviews.size();i++)
        {
            if(i==0)
            {
                sheet_names[0] = "Sheet contains following views: \n" + dfviews[i]->Name();
            }
            else
            {
                sheet_names[0] =  sheet_names[0] +"\n"+ dfviews[i]->Name();
                
            }
        }
        
        //Displaying the Drawing sheet view names
        theUI->NXMessageBox()->Show(sheet_name.GetText(), NXOpen::NXMessageBox::DialogTypeInformation,sheet_names[0].GetText());
        
        // ----------------------------------------------
        //   Menu: Edit->Delete...
        // ----------------------------------------------
        theSession->UpdateManager()->ClearErrorList();
        Session::UndoMarkId markId1;
        markId1 = theSession->SetUndoMark(Session::MarkVisibilityVisible, "Delete");
        
        int err1;
        err1 = theSession->UpdateManager()->AddToDeleteList(note1);
        
        int err2;
        err2= theSession->UpdateManager()->DoUpdate(markId1);
        
        
    }
    
    /* Terminate the API environment */
    errorcode = UF_terminate();
}

/*****************************************************************************
 **  Utilities
 *****************************************************************************/

/* Unload Handler
 **     This function specifies when to unload your application from Unigraphics.
 **     If your application registers a callback (from a MenuScript item or a
 **     User Defined Object for example), this function MUST return
 **     "UF_UNLOAD_UG_TERMINATE". */
extern int ufusr_ask_unload( void )
{
    return( UF_UNLOAD_IMMEDIATELY );
}

